// npmvet/dist/renderers/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var deps_1 = require("../deps");
exports.render = function (depMap) {
    /*
     * Merge dependencies and dev dependencies
     */
    var allPackages = depMap.deps.concat(depMap.devDeps);
    /*
     * Generate match info for each package
     */
    var packageMatchInfos = createMatchInfos(allPackages);
    /*
     * Generate indented JSON string
     */
    var output = JSON.stringify(packageMatchInfos, undefined, 2);
    /*
     * Print the JSON string
     */
    console.log(output);
};
/**
 * @param  {IPackageDescriptor[]} pkgDescriptor
 * @returns {IPackageMatchInfo[]}
 */
var createMatchInfos = function (pkgDescriptor) {
    return pkgDescriptor.map(function (pkg) { return ({
        name: pkg.name,
        packageVersion: pkg.parsedDefinedVersion,
        installedVersion: pkg.installedVersion,
        matches: deps_1.isMatchingVersion(pkg.parsedDefinedVersion, pkg.installedVersion),
        locked: pkg.locked
    }); });
};
